File: python/ssi_core/edge_common.py
from .scrambling import scr_to_scrambled_state_edg

EDGE_FLOAT_BUFFER_ORDER_PSEUDOSWAP = ('UF', 'UB', 'UR', 'UL', 'FR', 'FL', 'DF', 'DB', 'DR', 'DL')
EDGE_FLOAT_BUFFER_ORDER_WEAKSWAP = ('UF', 'UR', 'UB', 'UL', 'FR', 'FL', 'DF', 'DB', 'DR', 'DL')
STICKER_LETTER_MAP = {
    'UBL': 'C', 'UB': 'C',
    'UBR': 'B', 'UR': 'B',
    'UFR': 'Q', 'UF': 'Q',
    'UFL': 'D', 'UL': 'D',
    'LUB': 'E', 'LU': 'E',
    'LUF': 'F', 'LF': 'F',
    'LDF': 'G', 'LD': 'G',
    'LDB': 'H', 'LB': 'H',
    'FUL': 'I', 'RU': 'I',
    'FUR': 'Y', 'FU': 'Y',
    'FDR': 'M', 'FD': 'M',
    'FDL': 'L', 'FL': 'L',
    'FR': 'J',
    'RUF': 'W',
    'RUB': 'N', 'RB': 'N',
    'RDB': 'O', 'RD': 'O',
    'RDF': 'A', 'RF': 'A',
    'BUR': 'P', 'BU': 'P',
    'BUL': 'R', 'BL': 'R',
    'BDL': 'S', 'BD': 'S',
    'BDR': 'T', 'BR': 'T',
    'DFL': 'K', 'DF': 'K',
    'DFR': 'U', 'DR': 'U',
    'DBR': 'V', 'DB': 'V',
    'DBL': 'Z', 'DL': 'Z',
}

def remove_from_list_edg(visiting_list, sticker):
    new_list = visiting_list.copy()
    new_list.remove(sticker)
    new_list.remove(sticker[::-1])
    return new_list

def get_piece_group_edg(sticker):
    return [sticker, sticker[::-1]]

def normalize_edge_buffers(edge_buffers, edge_method):
    buffer_order = EDGE_FLOAT_BUFFER_ORDER_WEAKSWAP if edge_method == 'weakswap' else EDGE_FLOAT_BUFFER_ORDER_PSEUDOSWAP
    if edge_buffers == 'all':
        return list(buffer_order)
    if edge_buffers is None:
        return ['UF']
    allowed_buffers = set(edge_buffers)
    normalized_buffers = [buffer for buffer in buffer_order if buffer in allowed_buffers]
    if not normalized_buffers:
        assert False
    return normalized_buffers

def switch_with_buffer_edg(buffer, target, state):
    new_state = state.copy()
    buffer_group = get_piece_group_edg(buffer)
    target_group = get_piece_group_edg(target)
    new_state[buffer_group[0]] = state[target_group[0]]
    new_state[buffer_group[1]] = state[target_group[1]]
    new_state[target_group[0]] = state[buffer_group[0]]
    new_state[target_group[1]] = state[buffer_group[1]]
    return new_state

def apply_trace_log_edg(state, trace_log):
    virtual_state = state.copy()
    for buffer, target in trace_log:
        virtual_state = switch_with_buffer_edg(buffer, target, virtual_state)
    return virtual_state

def piece_in_its_place_edg(state, sticker):
    return state[sticker] in get_piece_group_edg(sticker)

def piece_solved_edg(state, sticker):
    return state[sticker] == sticker

def edge_buffer_solved_for_float(state, buffer, edge_method, parity=False):
    if edge_method == 'weakswap':
        if buffer == 'UF':
            return state['UF'] in ('UF', 'UR')
        if buffer == 'UR':
            return state['UR'] in ('UR', 'UF')
        return piece_solved_edg(state, buffer)
    if buffer == 'UF':
        return state['UF'] == ('UR' if parity else 'UF')
    if buffer == 'UR':
        return state['UR'] == ('UF' if parity else 'UR')
    return piece_solved_edg(state, buffer)

def flatten_edge_trace_segments(segments):
    return [target for segment in segments for target in segment['targets']]

def sticker_to_letter(sticker):
    return STICKER_LETTER_MAP[sticker]

def stickers_to_letters(stickers):
    return [sticker_to_letter(sticker) for sticker in stickers]

def segments_to_letter_view(segments):
    return [{'buffer': sticker_to_letter(segment['buffer']), 'targets': stickers_to_letters(segment['targets'])} for segment in segments]

def pair_letters(letters):
    return [''.join(letters[i:i + 2]) for i in range(0, len(letters), 2)]

def analyze_edge_trace_segments(segments):
    odd_segments = [segment for segment in segments if len(segment['targets']) % 2]
    even_segments = [segment for segment in segments if not len(segment['targets']) % 2]
    standalone_algs = sum((len(segment['targets']) + 1) // 2 for segment in segments)
    floating_algs = standalone_algs - (len(odd_segments) // 2)
    return {
        'odd_segments': odd_segments,
        'even_segments': even_segments,
        'edge_parity': bool(len(odd_segments) % 2),
        'algs': floating_algs,
    }

def trace_state_edg_floating(state, solved_list, flipped_list, edge_buffers, edge_method, parity=False):
    active_buffers = normalize_edge_buffers(edge_buffers, edge_method)
    current_buffer_i = 0
    current_buffer = active_buffers[current_buffer_i]
    all_edge_stickers = [
        'UF', 'FU', 'UR', 'RU', 'UB', 'BU', 'UL', 'LU', 'FR', 'RF', 'FL', 'LF',
        'DF', 'FD', 'DR', 'RD', 'DB', 'BD', 'BR', 'RB', 'BL', 'LB', 'DL', 'LD',
    ]
    need_visiting = [sticker for sticker in all_edge_stickers if sticker not in get_piece_group_edg(current_buffer)]
    for sticker in solved_list:
        need_visiting = remove_from_list_edg(need_visiting, sticker)
    for sticker in flipped_list:
        need_visiting = remove_from_list_edg(need_visiting, sticker)

    trace_log = []
    segments = []

    def ensure_current_segment():
        if not segments or segments[-1]['buffer'] != current_buffer:
            segments.append({'buffer': current_buffer, 'targets': []})

    def remove_if_needed(stickers):
        for sticker in stickers:
            if sticker in need_visiting:
                need_visiting.remove(sticker)

    while need_visiting:
        virtual_state = apply_trace_log_edg(state, trace_log)

        if (
            edge_method == 'weakswap'
            and current_buffer == 'UF'
            and set(need_visiting) == {'UR', 'RU'}
            and not (len(trace_log) % 2)
        ):
            flipped_list.append('UR')
            break

        if (
            edge_method == 'weakswap'
            and current_buffer == 'UF'
            and ('UR' in need_visiting or 'RU' in need_visiting)
            and virtual_state['UF'] in ('UF', 'FU', 'UR', 'RU')
        ):
            if virtual_state['UF'] in ('UR', 'RU'):
                target = virtual_state['UF']
            elif virtual_state['UF'] == 'UF':
                target = 'UR'
            else:
                target = 'RU'
            need_visiting = remove_from_list_edg(need_visiting, 'UR')
            ensure_current_segment()
            segments[-1]['targets'].append(target)
            trace_log.append((current_buffer, target))
            continue

        if (
            edge_method == 'pseudoswap'
            and current_buffer == 'UF'
            and ('UR' in need_visiting or 'RU' in need_visiting)
            and (
                (not parity and virtual_state['UF'] in ('UR', 'RU'))
                or (parity and virtual_state['UF'] in ('UF', 'FU'))
            )
        ):
            target = virtual_state['UF'].replace('F', 'R')
            need_visiting = remove_from_list_edg(need_visiting, 'UR')
            ensure_current_segment()
            segments[-1]['targets'].append(target)
            trace_log.append((current_buffer, target))
            continue

        solved_for_float = edge_buffer_solved_for_float(virtual_state, current_buffer, edge_method, parity=parity)
        if (
            edge_method == 'pseudoswap'
            and current_buffer == 'UF'
            and ('UR' in need_visiting or 'RU' in need_visiting)
        ):
            solved_for_float = False

        cycle_break_only = piece_in_its_place_edg(virtual_state, current_buffer) and not solved_for_float
        target = None
        if not (solved_for_float or cycle_break_only):
            candidate_target = virtual_state[current_buffer]
            if candidate_target not in need_visiting:
                cycle_break_only = True
            else:
                target = candidate_target
                need_visiting = remove_from_list_edg(need_visiting, target)

        if solved_for_float:
            floated = False
            for next_buffer_i in range(current_buffer_i + 1, len(active_buffers)):
                next_buffer = active_buffers[next_buffer_i]
                if edge_buffer_solved_for_float(virtual_state, next_buffer, edge_method, parity=parity):
                    continue
                current_buffer_i = next_buffer_i
                current_buffer = next_buffer
                remove_if_needed(get_piece_group_edg(current_buffer))
                floated = True
                break

            if floated:
                continue

            target = need_visiting[0]
        elif cycle_break_only:
            target = need_visiting[0]

        ensure_current_segment()
        segments[-1]['targets'].append(target)
        trace_log.append((current_buffer, target))

    return segments

__all__ = [
    'analyze_edge_trace_segments',
    'apply_trace_log_edg',
    'edge_buffer_solved_for_float',
    'flatten_edge_trace_segments',
    'get_piece_group_edg',
    'normalize_edge_buffers',
    'pair_letters',
    'piece_in_its_place_edg',
    'piece_solved_edg',
    'remove_from_list_edg',
    'scr_to_scrambled_state_edg',
    'segments_to_letter_view',
    'sticker_to_letter',
    'stickers_to_letters',
    'switch_with_buffer_edg',
    'trace_state_edg_floating',
]

File: web/edge-common.js
(function (global) {
  const deps = typeof module !== 'undefined' && module.exports
    ? require('./scrambling')
    : global.SsiCoreModules;

  const { scrToScrambledStateEdg } = deps;

  const EDGE_FLOAT_BUFFER_ORDER_PSEUDOSWAP = ['UF', 'UB', 'UR', 'UL', 'FR', 'FL', 'DF', 'DB', 'DR', 'DL'];
  const EDGE_FLOAT_BUFFER_ORDER_WEAKSWAP = ['UF', 'UR', 'UB', 'UL', 'FR', 'FL', 'DF', 'DB', 'DR', 'DL'];
  const STICKER_LETTER_MAP = {
    UBL: 'C', UB: 'C', UBR: 'B', UR: 'B', UFR: 'Q', UF: 'Q', UFL: 'D', UL: 'D',
    LUB: 'E', LU: 'E', LUF: 'F', LF: 'F', LDF: 'G', LD: 'G', LDB: 'H', LB: 'H',
    FUL: 'I', RU: 'I', FUR: 'Y', FU: 'Y', FDR: 'M', FD: 'M', FDL: 'L', FL: 'L', FR: 'J',
    RUF: 'W', RUB: 'N', RB: 'N', RDB: 'O', RD: 'O', RDF: 'A', RF: 'A',
    BUR: 'P', BU: 'P', BUL: 'R', BL: 'R', BDL: 'S', BD: 'S', BDR: 'T', BR: 'T',
    DFL: 'K', DF: 'K', DFR: 'U', DR: 'U', DBR: 'V', DB: 'V', DBL: 'Z', DL: 'Z',
  };

  function removeFromListEdg(visitingList, sticker) {
    const newList = [...visitingList];
    newList.splice(newList.indexOf(sticker), 1);
    newList.splice(newList.indexOf(sticker.split('').reverse().join('')), 1);
    return newList;
  }

  function getPieceGroupEdg(sticker) {
    return [sticker, sticker.split('').reverse().join('')];
  }

  function normalizeEdgeBuffers(edgeBuffers, edgeMethod) {
    const bufferOrder = edgeMethod === 'weakswap' ? EDGE_FLOAT_BUFFER_ORDER_WEAKSWAP : EDGE_FLOAT_BUFFER_ORDER_PSEUDOSWAP;
    if (edgeBuffers === 'all') return [...bufferOrder];
    if (edgeBuffers == null) return ['UF'];
    const allowedBuffers = new Set(edgeBuffers);
    const normalized = bufferOrder.filter((buffer) => allowedBuffers.has(buffer));
    if (!normalized.length) throw new Error('Assertion failed');
    return normalized;
  }

  function switchWithBufferEdg(buffer, target, state) {
    const newState = { ...state };
    const bufferGroup = getPieceGroupEdg(buffer);
    const targetGroup = getPieceGroupEdg(target);
    newState[bufferGroup[0]] = state[targetGroup[0]];
    newState[bufferGroup[1]] = state[targetGroup[1]];
    newState[targetGroup[0]] = state[bufferGroup[0]];
    newState[targetGroup[1]] = state[bufferGroup[1]];
    return newState;
  }

  function applyTraceLogEdg(state, traceLog) {
    let virtualState = { ...state };
    for (const [buffer, target] of traceLog) {
      virtualState = switchWithBufferEdg(buffer, target, virtualState);
    }
    return virtualState;
  }

  function pieceInItsPlaceEdg(state, sticker) {
    return getPieceGroupEdg(sticker).includes(state[sticker]);
  }

  function pieceSolvedEdg(state, sticker) {
    return state[sticker] === sticker;
  }

  function edgeBufferSolvedForFloat(state, buffer, edgeMethod, parity = false) {
    if (edgeMethod === 'weakswap') {
      if (buffer === 'UF') return ['UF', 'UR'].includes(state.UF);
      if (buffer === 'UR') return ['UR', 'UF'].includes(state.UR);
      return pieceSolvedEdg(state, buffer);
    }
    if (buffer === 'UF') return state.UF === (parity ? 'UR' : 'UF');
    if (buffer === 'UR') return state.UR === (parity ? 'UF' : 'UR');
    return pieceSolvedEdg(state, buffer);
  }

  function flattenEdgeTraceSegments(segments) {
    return segments.flatMap((segment) => segment.targets);
  }

  function stickerToLetter(sticker) {
    return STICKER_LETTER_MAP[sticker];
  }

  function stickersToLetters(stickers) {
    return stickers.map(stickerToLetter);
  }

  function segmentsToLetterView(segments) {
    return segments.map((segment) => ({ buffer: stickerToLetter(segment.buffer), targets: stickersToLetters(segment.targets) }));
  }

  function pairLetters(letters) {
    const pairs = [];
    for (let i = 0; i < letters.length; i += 2) pairs.push(letters.slice(i, i + 2).join(''));
    return pairs;
  }

  function analyzeEdgeTraceSegments(segments) {
    const oddSegments = segments.filter((segment) => segment.targets.length % 2);
    const evenSegments = segments.filter((segment) => !(segment.targets.length % 2));
    const standaloneAlgs = segments.reduce((sum, segment) => sum + Math.floor((segment.targets.length + 1) / 2), 0);
    const floatingAlgs = standaloneAlgs - Math.floor(oddSegments.length / 2);
    return {
      odd_segments: oddSegments,
      even_segments: evenSegments,
      edge_parity: Boolean(oddSegments.length % 2),
      algs: floatingAlgs,
    };
  }

  function traceStateEdgFloating(state, solvedList, flippedList, edgeBuffers, edgeMethod, parity = false) {
    const activeBuffers = normalizeEdgeBuffers(edgeBuffers, edgeMethod);
    let currentBufferI = 0;
    let currentBuffer = activeBuffers[currentBufferI];
    let needVisiting = [
      'UF', 'FU', 'UR', 'RU', 'UB', 'BU', 'UL', 'LU', 'FR', 'RF', 'FL', 'LF',
      'DF', 'FD', 'DR', 'RD', 'DB', 'BD', 'BR', 'RB', 'BL', 'LB', 'DL', 'LD',
    ].filter((sticker) => !getPieceGroupEdg(currentBuffer).includes(sticker));
    for (const sticker of solvedList) needVisiting = removeFromListEdg(needVisiting, sticker);
    for (const sticker of flippedList) needVisiting = removeFromListEdg(needVisiting, sticker);

    const traceLog = [];
    const segments = [];

    const ensureCurrentSegment = () => {
      if (!segments.length || segments[segments.length - 1].buffer !== currentBuffer) {
        segments.push({ buffer: currentBuffer, targets: [] });
      }
    };

    const removeIfNeeded = (stickers) => {
      for (const sticker of stickers) {
        const idx = needVisiting.indexOf(sticker);
        if (idx !== -1) needVisiting.splice(idx, 1);
      }
    };

    while (needVisiting.length) {
      const virtualState = applyTraceLogEdg(state, traceLog);

      if (
        edgeMethod === 'weakswap'
        && currentBuffer === 'UF'
        && new Set(needVisiting).size === 2
        && needVisiting.includes('UR')
        && needVisiting.includes('RU')
        && !(traceLog.length % 2)
      ) {
        flippedList.push('UR');
        break;
      }

      if (
        edgeMethod === 'weakswap'
        && currentBuffer === 'UF'
        && (needVisiting.includes('UR') || needVisiting.includes('RU'))
        && ['UF', 'FU', 'UR', 'RU'].includes(virtualState.UF)
      ) {
        let target;
        if (['UR', 'RU'].includes(virtualState.UF)) target = virtualState.UF;
        else if (virtualState.UF === 'UF') target = 'UR';
        else target = 'RU';
        needVisiting = removeFromListEdg(needVisiting, 'UR');
        ensureCurrentSegment();
        segments[segments.length - 1].targets.push(target);
        traceLog.push([currentBuffer, target]);
        continue;
      }

      if (
        edgeMethod === 'pseudoswap'
        && currentBuffer === 'UF'
        && (needVisiting.includes('UR') || needVisiting.includes('RU'))
        && ((!parity && ['UR', 'RU'].includes(virtualState.UF)) || (parity && ['UF', 'FU'].includes(virtualState.UF)))
      ) {
        const target = virtualState.UF.replace('F', 'R');
        needVisiting = removeFromListEdg(needVisiting, 'UR');
        ensureCurrentSegment();
        segments[segments.length - 1].targets.push(target);
        traceLog.push([currentBuffer, target]);
        continue;
      }

      let solvedForFloat = edgeBufferSolvedForFloat(virtualState, currentBuffer, edgeMethod, parity);
      if (edgeMethod === 'pseudoswap' && currentBuffer === 'UF' && (needVisiting.includes('UR') || needVisiting.includes('RU'))) {
        solvedForFloat = false;
      }

      let cycleBreakOnly = pieceInItsPlaceEdg(virtualState, currentBuffer) && !solvedForFloat;
      let target = null;
      if (!(solvedForFloat || cycleBreakOnly)) {
        const candidateTarget = virtualState[currentBuffer];
        if (!needVisiting.includes(candidateTarget)) cycleBreakOnly = true;
        else {
          target = candidateTarget;
          needVisiting = removeFromListEdg(needVisiting, target);
        }
      }

      if (solvedForFloat) {
        let floated = false;
        for (let nextBufferI = currentBufferI + 1; nextBufferI < activeBuffers.length; nextBufferI += 1) {
          const nextBuffer = activeBuffers[nextBufferI];
          if (edgeBufferSolvedForFloat(virtualState, nextBuffer, edgeMethod, parity)) continue;
          currentBufferI = nextBufferI;
          currentBuffer = nextBuffer;
          removeIfNeeded(getPieceGroupEdg(currentBuffer));
          floated = true;
          break;
        }
        if (floated) continue;
        target = needVisiting[0];
      } else if (cycleBreakOnly) {
        target = needVisiting[0];
      }

      ensureCurrentSegment();
      segments[segments.length - 1].targets.push(target);
      traceLog.push([currentBuffer, target]);
    }

    return segments;
  }

  const api = {
    EDGE_FLOAT_BUFFER_ORDER_PSEUDOSWAP,
    EDGE_FLOAT_BUFFER_ORDER_WEAKSWAP,
    STICKER_LETTER_MAP,
    analyzeEdgeTraceSegments,
    applyTraceLogEdg,
    edgeBufferSolvedForFloat,
    flattenEdgeTraceSegments,
    getPieceGroupEdg,
    normalizeEdgeBuffers,
    pairLetters,
    pieceInItsPlaceEdg,
    pieceSolvedEdg,
    removeFromListEdg,
    scrToScrambledStateEdg,
    segmentsToLetterView,
    stickerToLetter,
    stickersToLetters,
    switchWithBufferEdg,
    traceStateEdgFloating,
  };

  global.SsiCoreModules = global.SsiCoreModules || {};
  Object.assign(global.SsiCoreModules, api);

  if (typeof module !== 'undefined' && module.exports) {
    module.exports = api;
  }
})(typeof globalThis !== 'undefined' ? globalThis : this);
